server/server.go

```diff
@@ -166,90 +172,94 @@ func Start() (err error) {
+		gaia.Cfg.Logger.Warn("using auto-generated key to sign jwt tokens, do not use in production")
+		jwtKey = make([]byte, 64)
+		_, err = rand.Read(jwtKey.([]byte))
+		if err != nil {
+			gaia.Cfg.Logger.Error("error auto-generating jwt key", "error", err.Error())
+			return
 		}
-		v, err := services.VaultService(nil)
+	} else {
+		keyData, err := ioutil.ReadFile(gaia.Cfg.JwtPrivateKeyPath)
 		if err != nil {
-			gaia.Cfg.Logger.Error("error initiating vault")
+			gaia.Cfg.Logger.Error("could not read jwt key file", "error", err.Error())
 			return err
 		}
-		if err = v.LoadSecrets(); err != nil {
-			gaia.Cfg.Logger.Error("error loading secrets from vault")
+		jwtKey, err = jwt.ParseRSAPrivateKeyFromPEM(keyData)
+		if err != nil {
+			gaia.Cfg.Logger.Error("could not parse jwt key file", "error", err.Error())
 			return err
 		}
+	}
+	gaia.Cfg.JWTKey = jwtKey
 
-		// Generate global worker secret if it does not exist
-		_, err = v.Get(gaia.WorkerRegisterKey)
-		if err != nil {
-			// Secret hasn't been generated yet
-			gaia.Cfg.Logger.Info("global worker registration secret has not been generated yet. Will generate it now...")
-			secret := []byte(security.GenerateRandomUUIDV5())
-
-			// Store secret in vault
-			v.Add(gaia.WorkerRegisterKey, secret)
-			if err := v.SaveSecrets(); err != nil {
-				gaia.Cfg.Logger.Error("failed to store secret into vault", "error", err.Error())
-				return err
-			}
-		}
+	// Initialize echo instance
+	echoInstance = echo.New()
 
-		// Initialize handlers
-		err = handlers.InitHandlers(echoInstance)
-		if err != nil {
-			gaia.Cfg.Logger.Error("cannot initialize handlers", "error", err.Error())
+	// Initiating Vault
+	if gaia.Cfg.VaultPath == "" {
+		// Set default to data folder
+		gaia.Cfg.VaultPath = gaia.Cfg.DataPath
+	}
+	v, err := services.VaultService(nil)
+	if err != nil {
+		gaia.Cfg.Logger.Error("error initiating vault")
+		return err
+	}
+	if err = v.LoadSecrets(); err != nil {
+		gaia.Cfg.Logger.Error("error loading secrets from vault")
+		return err
+	}
+
+	// Generate global worker secret if it does not exist
+	_, err = v.Get(gaia.WorkerRegisterKey)
+	if err != nil {
+		// Secret hasn't been generated yet
+		gaia.Cfg.Logger.Info("global worker registration secret has not been generated yet. Will generate it now...")
+		secret := []byte(security.GenerateRandomUUIDV5())
+
+		// Store secret in vault
+		v.Add(gaia.WorkerRegisterKey, secret)
+		if err := v.SaveSecrets(); err != nil {
+			gaia.Cfg.Logger.Error("failed to store secret into vault", "error", err.Error())
 			return err
 		}
 	}
 
+	// Initialize handlers
+	err = handlers.InitHandlers(echoInstance)
+	if err != nil {
+		gaia.Cfg.Logger.Error("cannot initialize handlers", "error", err.Error())
+		return err
+	}
+
 	// Initialize scheduler
 	scheduler, err := services.SchedulerService()
 	if err != nil {
 		return
 	}
 
+	// Start worker gRPC server.
+	// We need this in both modes (server and worker) for docker worker to run.
```

Comment:
What's the reason for needing it in both?

Reply:
Good question. Imagine you want to run pipelines on a remote Gaia worker. In this case the following will happen:

1. Worker retrieves pipeline run from Gaia primary instance.
2. Worker starts another worker in a docker container.
3. Worker in the docker container connects to the worker.
4. Worker in docker container executes pipeline.
5. Worker in docker container sends back run logs & status.
6. Worker sends back logs & status to primary instance.

Reply:
Ahh I see. Makes sense.